internal/xds/rbac/rbac_engine_test.go

```diff
@@ -271,336 +428,514 @@ func (s) TestRBACEngineConstruction(t *testing.T) {
+				// shouldn't be allowed to proceed as didn't match with allow.
+				{
+					rpcData: &rpcData{
+						destinationPort: 8000,
+						peerInfo: &peer.Peer{
+							Addr: &addr{ipAddress: "10.0.0.0"},
+						},
 					},
-					wantMatchingPolicyName: "",
+					wantStatusCode: codes.PermissionDenied,
+				},
+				// This RPC shouldn't match with allow, match with deny, and
+				// thus shouldn't be allowed to proceed.
+				{
+					rpcData: &rpcData{
+						md: map[string][]string{
+							":path": {"localhost-fan-page"},
+						},
+						destinationPort: 8080,
+						peerInfo: &peer.Peer{
+							Addr: &addr{ipAddress: "10.0.0.0"},
+						},
+					},
+					wantStatusCode: codes.PermissionDenied,
 				},
 			},
 		},
 	}
+
 	for _, test := range tests {
 		t.Run(test.name, func(t *testing.T) {
-			// Instantiate the rbacEngine with different configurations that
+			// Instantiate the chainedRBACEngine with different configurations that are
 			// interesting to test and to query.
-			rbacEngine, err := NewEngine(test.rbacConfig)
+			cre, err := NewChainEngine(test.rbacConfigs)
 			if err != nil {
 				t.Fatalf("Error constructing RBAC Engine: %v", err)
 			}
-			// Query that created RBAC Engine with different args to see if the
-			// RBAC Engine configured a certain way works as intended.
-			for _, queryToRBACEngine := range test.rbacQueries {
-				// The matchingPolicyName returned will be empty in the case of
-				// no matching policy. Thus, matchingPolicyName can also be used
-				// to test the "error" condition of no matching policies.
-				matchingPolicyName, matchingPolicyFound := rbacEngine.FindMatchingPolicy(queryToRBACEngine.rpcData)
-				if matchingPolicyFound != queryToRBACEngine.wantMatch || matchingPolicyName != queryToRBACEngine.wantMatchingPolicyName {
-					t.Errorf("FindMatchingPolicy(%+v) returned (%v, %v), want (%v, %v)", queryToRBACEngine.rpcData, matchingPolicyName, matchingPolicyFound, queryToRBACEngine.wantMatchingPolicyName, queryToRBACEngine.wantMatch)
+			// Query the created chain of RBAC Engines with different args to see
+			// if the chain of RBAC Engines configured as such works as intended.
+			for _, data := range test.rbacQueries {
+				// Construct the context with three data points that have enough
+				// information to represent incoming RPC's. This will be how a
+				// user uses this API. A user will have to put MD, PeerInfo, and
+				// the connection the RPC is sent on in the context.
+				ctx := metadata.NewIncomingContext(context.Background(), data.rpcData.md)
+
+				// Make a TCP connection with a certain destination port. The
+				// address/port of this connection will be used to populate the
+				// destination ip/port in RPCData struct. This represents what
+				// the user of ChainEngine will have to place into
+				// context, as this is only way to get destination ip and port.
+				lis, err := net.Listen("tcp", "localhost:"+fmt.Sprint(data.rpcData.destinationPort))
```

Comment:
You cannot prescribe a port to listen on for a real listener.  Use a fake listener or let it find a free port on the machine.

Reply:
I agree, but all this is doing is providing a way to create a connection with a certain dest ip/port for testing purposes (where we can specify the port as well). Letting it find a free port on the machine, dialing to it, and then using that connection would lead to nondeterministic and flaky tests. Do you see another way of testing dest ip/port?

Reply:
You probably need to set up your RBAC rules _after_ you set up your listener, then.  You simply cannot pick the port ahead of time, as it may not actually be available when the tests are running.  It might pass on GA but fail locally/elsewhere.

Reply:
+1
Sorry, didn't notice this earlier.

Reply:
Deleted logic related to destination Port, and had it listen on the wildcard port 0.